source/h1_encoder.c

```diff
@@ -319,6 +344,309 @@ int aws_h1_encoder_start_message(
+    return aws_byte_buf_write_from_whole_cursor(dst, crlf_cursor);
+}
+
+bool write_chunk_size(struct aws_byte_buf *dst, struct aws_input_stream *chunk) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION(chunk);
+    AWS_PRECONDITION((dst->capacity - dst->len) >= MAX_ASCII_HEX_CHUNK_STR_SIZE);
+    int64_t chunk_size = 0;
+    if (AWS_OP_SUCCESS != aws_input_stream_get_length(chunk, &chunk_size)) {
+        return false;
+    }
+    char ascii_hex_chunk_size_str[MAX_ASCII_HEX_CHUNK_STR_SIZE];
+    AWS_ZERO_ARRAY(ascii_hex_chunk_size_str);
+    snprintf(ascii_hex_chunk_size_str, sizeof(ascii_hex_chunk_size_str), "%zX", chunk_size);
+    return aws_byte_buf_write_from_whole_cursor(dst, aws_byte_cursor_from_c_str(ascii_hex_chunk_size_str));
+}
+
+bool write_chunk_extension(struct aws_byte_buf *dst, struct aws_http1_chunk_extension *chunk_extension) {
+    AWS_PRECONDITION(chunk_extension);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    bool wrote_all = true;
+    wrote_all &= aws_byte_buf_write_u8(dst, ';');
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(dst, chunk_extension->key);
+    wrote_all &= aws_byte_buf_write_u8(dst, '=');
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(dst, chunk_extension->value);
+    return wrote_all;
+}
+
+static bool s_fill_byte_buffer(struct aws_byte_buf *dst, struct aws_byte_cursor *src) {
+    size_t write_size = aws_min_size(dst->capacity - dst->len, src->len);
+    struct aws_byte_cursor sub_cursor;
+    sub_cursor.len = write_size;
+    sub_cursor.ptr = src->ptr;
+    if (AWS_UNLIKELY(AWS_OP_SUCCESS != aws_byte_buf_append(dst, &sub_cursor))) {
+        return false;
+    }
+    aws_byte_cursor_advance(src, write_size);
+    return 0 == src->len;
+}
+
+static bool s_chunk_line_state(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(encoder);
+    AWS_PRECONDITION(encoder->message);
+    AWS_PRECONDITION(encoder->message->body_chunk);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION(aws_byte_buf_is_valid(&encoder->message->body_chunk->chunk_line));
+    bool completely_written = s_fill_byte_buffer(dst, &encoder->message->body_chunk->chunk_line_cursor);
+    if (completely_written) {
+        aws_byte_buf_clean_up(&encoder->message->body_chunk->chunk_line);
+        return true;
+    }
+    return false;
+}
+
+static bool s_chunk_payload_state(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder, int *aws_op_result) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION(encoder);
+    AWS_PRECONDITION(encoder->message);
+    AWS_PRECONDITION(encoder->message->body_chunk);
+    AWS_PRECONDITION(encoder->message->body_chunk->data);
+    AWS_PRECONDITION(aws_op_result);
+
+    size_t prev_len = dst->len;
+    if (AWS_UNLIKELY(aws_input_stream_read(encoder->message->body_chunk->data, dst))) {
```

Comment:
are we checking that the we don't stream more or less than the size we wrote at the start of the chunk?

fwiw, I did like the aws_input_stream_read_n() function you'd added once upon a time

Reply:
I had not been checking it as the chunk size was taken from the input stream itself and would have been an internal API bug. Once the API is updated to expose an explicit size call, I am checking using the progress_bytes in the encoder to calculate the actual payload size written (excluding chunked encoding overhead and anything on the chunk line). Let me know if progress_bytes is the wrong place to keep that information.